drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2252 @@
+
+	/* block count enable */
+	rcar_mmc_write_reg32(dev, RCAR_MMC_STOP, RCAR_MMC_STOP_SEC);
+	/* number of transfer blocks */
+	rcar_mmc_write_reg32(dev, RCAR_MMC_SECCNT, 0);
+
+	/*
+	 * SD_BUF0 data swap disabled.
+	 * Read/write access to SD_BUF0 can be performed with the 64-bit access.
+	 *
+	 * Note: when using the DMA, the bus width should be fixed at 64 bits.
+	 */
+	rcar_mmc_write_reg32(dev, RCAR_MMC_HOST_MODE, 0);
+	data->width_access_sd_buf0 = 8;
+
+	/* disable sampling clock controller, it is used for uhs/sdr104, hs200 and hs400 */
+	ret = rcar_mmc_disable_scc(dev);
+	if (ret) {
+		return ret;
+	}
+
+	/*
+	 * configure divider inside MMC controller
+	 * set maximum possible divider
+	 */
+	ios.clock = data->props.f_min;
+	rcar_mmc_set_clk_rate(dev, &ios);
+
+	data->restore_cfg_after_reset = 1;
+
+	return 0;
+}
+
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+static void rcar_mmc_irq_handler(const void *arg)
+{
+	const struct device *dev = arg;
+
+	uint32_t dma_info1 = rcar_mmc_read_reg32(dev, RCAR_MMC_DMA_INFO1);
+	uint32_t dma_info2 = rcar_mmc_read_reg32(dev, RCAR_MMC_DMA_INFO2);
+
+	if (dma_info1 || dma_info2) {
+		struct mmc_rcar_data *data = dev->data;
+
+		rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO1_MASK, 0xfffffeff);
+		rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO2_MASK, ~0);
+		k_sem_give(&data->irq_xref_fin);
+	} else {
+		LOG_WRN("%s: warning: non-dma event triggers irq", dev->name);
+	}
+}
+#endif /* CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT */
+
+/* initialize and configure the Renesas MMC driver */
+static int rcar_mmc_init(const struct device *dev)
+{
+	int ret = 0;
+	struct mmc_rcar_data *data = dev->data;
+	const struct mmc_rcar_cfg *cfg = dev->config;
+
+	if (unlikely(cfg == NULL)) {
+		LOG_ERR("%s: there isn't config inside device structure", dev->name);
+		return -EFAULT;
+	}
+
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+	ret = k_sem_init(&data->irq_xref_fin, 0, 1);
+	if (ret) {
+		LOG_ERR("%s: can't init semaphore", dev->name);
+		return ret;
+	}
+#endif
+
+	if (unlikely(dev->data == NULL)) {
+		LOG_ERR("%s: there isn't data instance inside device structure", dev->name);
+		return -EFAULT;
+	}
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("%s: error can't apply pinctrl state", dev->name);
+		goto exit_unmap;
+	}
+
+	if (!device_is_ready(cfg->cpg_dev)) {
+		LOG_ERR("%s: error cpg_dev isn't ready", dev->name);
+		ret = -ENODEV;
+		goto exit_unmap;
+	}
+
+	ret = rcar_mmc_init_start_clk(cfg);
+	if (ret < 0) {
+		LOG_ERR("%s: error can't turn on the cpg", dev->name);
+		goto exit_unmap;
+	}
+
+	/* it's needed for SDHC */
+	rcar_mmc_init_host_props(dev);
+
+	/*
+	 * note: we need to set this flag before call rcar_mmc_init_controller_regs,
+	 *       because we want to cleanup dma regs and mask dma irqs
+	 */
+	data->dma_support = IS_ENABLED(CONFIG_RCAR_MMC_DMA_SUPPORT);
```

Comment:
Why does this flag need to be in the `data` structure? It seems like we could enable DMA support in code by using `if(IS_ENABLED(CONFIG_RCAR_MMC_DMA_SUPPORT))`, or at a minimum declare `dma_support` as part of the device `config` struct so the definition is not stored in RAM

Reply:
There was a plan to add some shell functionality and the possibility to disable DMA during runtime.

I'll use define instead